hoodie-common/src/main/java/com/uber/hoodie/common/HoodieFileGroupBuilder.java

```diff
@@ -0,0 +1,184 @@
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package com.uber.hoodie.common;
+
+import com.google.common.base.Preconditions;
+import com.uber.hoodie.common.model.FileSlice;
+import com.uber.hoodie.common.model.HoodieDataFile;
+import com.uber.hoodie.common.model.HoodieFileGroup;
+import com.uber.hoodie.common.model.HoodieLogFile;
+import com.uber.hoodie.common.table.HoodieTimeline;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.NavigableMap;
+import java.util.TreeMap;
+import org.apache.commons.lang3.builder.Builder;
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+
+/**
+ * FileGroupBuilder for Hoodie File Group.
+ * Handles cases when delta-commits happened after compaction requests.
+ * Ensures the latest file-slice for RT view contains all the log-files including
+ * the ones after pending compaction requests.
+ *
+ * File-Group/File-Slice Constraints maintained by this class:
+ * ===========================================================
+ * 1. A file-group is identified by an unique file-Id within a partition.
+ * 2. A file-group has one or more file-slices.
+ * 3. Exactly one file-slice for a file-group provides the latest snapshot view (A view which is capable of providing
+ *    all the latest records in that file-group).
+ * 4. A file-group can have at-most one pending compaction.
+ * 5. The latest file-slice in a file-group will include all of
+ *    (a) the data-file corresponding to the last committed instant before a compaction request instant (if present)
+ *    (b) the set of log-files due to delta-instants after the instant (a) but before compaction request instant
+ *    (c) the set of log-files written after compaction request instant
+ *
+ */
+public class HoodieFileGroupBuilder implements Builder<HoodieFileGroup> {
+
+  private static final transient Logger log = LogManager.getLogger(HoodieFileGroupBuilder.class);
+
+  /**
+   * Slices of files in this group, sorted with greater commit first.
+   */
+  private final NavigableMap<String, FileSlice> fileSlices = new TreeMap<>(HoodieFileGroup.getCommitTimeComparator());
+  /**
+   * Partition containing the file group.
+   */
+  private final String partitionPath;
+  /**
+   * uniquely identifies the file group
+   */
+  private final String id;
+  /**
+   * Timeline, based on which all getter work
+   */
+  private final HoodieTimeline timeline;
+
+  public HoodieFileGroupBuilder(String partitionPath, String id, HoodieTimeline timeline) {
+    this.partitionPath = partitionPath;
+    this.id = id;
+    this.timeline = timeline;
+  }
+
+  /**
+   * Add a new datafile into the file group
+   */
+  public HoodieFileGroupBuilder withDataFile(HoodieDataFile dataFile) {
```

Comment:
Should skipping of dataFiles etc happen here ? I think that's an operation that should happen at a higher level, say the TableFileSystem level.

Reply:
Hmm, that's fair.

Reply:
Hmm, I see the benefit and the reasons are fair. I still feel the FileGroup should just return whatever it can find on disk and then we apply filters on it to correctly return the file-groups, this is the way we do it right now too. At the very minimum, separation of concerns of like 1-1 mapping to what file-groups you find on disk and then filter, this way if there is a use case to delete orphan file groups etc we can do that too ? WDYT ?

Reply:
Good point on orphan files. I agree that all physical files should be trackable from FileGroup abstraction. The PR did have a bug in orphan file-groups handling. I have made changes to allow FileBuilder create files corresponding to inflight/orphan file-groups and added more test-cases to cover these cases.